src/commonMain/kotlin/app/cash/turbine/flow.kt

```diff
@@ -49,19 +80,59 @@ import kotlinx.coroutines.test.UnconfinedTestDispatcher
 public suspend fun <T> Flow<T>.test(
   timeout: Duration? = null,
   name: String? = null,
-  validate: suspend ReceiveTurbine<T>.() -> Unit,
+  validate: suspend TurbineTestContext<T>.() -> Unit,
 ) {
-  coroutineScope {
-    collectTurbineIn(this, null, name).apply {
-      if (timeout != null) {
-        withTurbineTimeout(timeout) {
-          validate()
+  val turbineRegistry = mutableListOf<ChannelTurbine<*>>()
+  reportTurbines(turbineRegistry) {
+    coroutineScope {
+      collectTurbineIn(this, null, name).apply {
+        try {
+          val testContext = TurbineTestContextImpl(this@apply, currentCoroutineContext())
+          if (timeout != null) {
+            withTurbineTimeout(timeout) {
+              testContext.validate()
+            }
+          } else {
+            testContext.validate()
+          }
+          cancel()
+          ensureAllEventsConsumed()
+        } catch (e: Throwable) {
+          // The exception needs to be reraised. However, if there are any unconsumed events
+          // from other turbines (including this one), those may indicate an underlying problem.
+          // So: create a report with all the registered turbines, and include exception as cause
+          val reportsWithUnconsumedEvents = turbineRegistry.map {
+            it.reportUnconsumedEvents()
+              // The exception will have cancelled its job hierarchy, producing cancellation exceptions
+              // in its wake. These aren't meaningful test feedback
+              .stripCancellations()
+          }
+            .filter { it.unconsumed.isNotEmpty() }
+          if (reportsWithUnconsumedEvents.isEmpty()) {
+            throw e
+          } else {
+            throw TurbineAssertionError(
+              buildString {
+                reportsWithUnconsumedEvents.forEach {
+                  it.describe(this@buildString)
+                  it.cause?.let { cause ->
+                    appendLine(
+                      """
+                        |
+                        |
+                        |Stack trace:
+                      """.trimMargin(),
+                    )
+                    append(cause.stackTraceToString())
+                    appendLine()
+                  }
+                }
+              },
+              e,
```

Comment:
This is the other part that feels odd. 

The new `TurbineAssertionError` here describes all the uncaught errors for all Turbines run within its scope, but buries the underlying exception that actually triggered this failure. This is often a good thing, but sometimes not. 

My first pass was for *any* unconsumed event to be reported here, but maybe this all makes more sense if everything but exceptions are stripped out.

Reply:
Ditched the idea of reporting any unconsumed event, switched to only reporting unconsumed errors. Works like a charm